src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -337,10 +349,69 @@ public String toCheckpoint(EntityModel model) {
             if (model.getThreshold() != null) {
                 json.addProperty(ENTITY_THRESHOLD, gson.toJson(model.getThreshold()));
             }
+            if (model.getErcf().isPresent()) {
+                json.addProperty(ENTITY_ERCF, toCheckpoint(model.getErcf().get()));
+            }
             return gson.toJson(json);
         });
     }
 
+    private String toCheckpoint(ExtendedRandomCutForest ercf) {
+        String checkpoint = null;
+        Map.Entry<LinkedBuffer, Boolean> result = checkoutOrNewBuffer();
+        LinkedBuffer buffer = result.getKey();
+        boolean requireCheckin = result.getValue();
+        try {
+            checkpoint = toCheckpoint(ercf, buffer);
+        } catch (RuntimeException e) {
+            logger.error("Failed to serialize model", e);
+            if (requireCheckin) {
+                try {
+                    serializeRCFBufferPool.invalidateObject(buffer);
+                    requireCheckin = false;
+                } catch (Exception x) {
+                    logger.warn("Failed to invalidate buffer", x);
+                }
+                checkpoint = toCheckpoint(ercf, LinkedBuffer.allocate(serializeRCFBufferSize));
+            }
+        } finally {
+            if (requireCheckin) {
+                try {
+                    serializeRCFBufferPool.returnObject(buffer);
+                } catch (Exception e) {
+                    logger.warn("Failed to return buffer to pool", e);
+                }
+            }
+        }
```

Comment:
[Non-blocking, can still approve PR] Since the logic inside the `finally` block is not intended to be run if the `catch` block is triggered, it would be simpler to just remove the `finally` block and move the logic in it into the `try` block.